src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1282 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Net.Http.HPack;
+using System.Net.Security;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+// Some stuff to do:
+
+// * Handle request body
+// * Handle request cancellation
+// * Flow control (both inbound and outbound)
+// * Various TODOs below
+
+namespace System.Net.Http
+{
+    internal sealed class Http2Connection : HttpConnectionBase, IDisposable
+    {
+        private readonly HttpConnectionPool _pool;
+        private readonly SslStream _stream;
+        private readonly object _syncObject;
+
+        // NOTE: These are mutable structs; do not make these readonly.
+        private ArrayBuffer _incomingBuffer;
+        private ArrayBuffer _outgoingBuffer;
+
+        private readonly HPackDecoder _hpackDecoder;
+
+        private readonly Dictionary<int, Http2Stream> _httpStreams;
+
+        private readonly SemaphoreSlim _writerLock;
+
+        private int _nextStream;
+        private bool _expectingSettingsAck;
+
+        private bool _disposed;
+
+        private const int MaxStreamId = int.MaxValue;
+
+        private static readonly byte[] s_http2ConnectionPreface = Encoding.ASCII.GetBytes("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n");
+
+        private const int InitialBufferSize = 4096;
+
+        public Http2Connection(HttpConnectionPool pool, SslStream stream)
+        {
+            _pool = pool;
+            _stream = stream;
+            _syncObject = new object();
+            _incomingBuffer = new ArrayBuffer(InitialBufferSize);
+            _outgoingBuffer = new ArrayBuffer(InitialBufferSize);
+
+            _hpackDecoder = new HPackDecoder();
+
+            _httpStreams = new Dictionary<int, Http2Stream>();
+
+            _writerLock = new SemaphoreSlim(1, 1);
+
+            _nextStream = 1;
+        }
+
+        public async Task SetupAsync()
+        {
+            // Send connection preface
+            _outgoingBuffer.EnsureAvailableSpace(s_http2ConnectionPreface.Length);
+            s_http2ConnectionPreface.AsSpan().CopyTo(_outgoingBuffer.AvailableSpan);
+            _outgoingBuffer.Commit(s_http2ConnectionPreface.Length);
+
+            // Send empty settings frame
+            WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.None, 0));
+
+            // TODO: We should disable PUSH_PROMISE here.
+            // TODO: We should send a connection-level WINDOW_UPDATE to allow
+            // a large amount of data to be received on the connection.  
+            // We don't care that much about connection-level flow control, we'll manage it per-stream.
```

Comment:
Will these sizes be configurable?

Reply:
Ideally, yes. Whether that happens for 2.2 or not, unclear.